hyeinkim/2021-04-15-Mockito 시작하기.md

```diff
@@ -0,0 +1,290 @@
+    void createStudyService() {
+        MemberService memberService = mock(MemberService.class);
+        StudyRepository studyRepository = mock(StudyRepository.class);
+        StudyService studyService = new StudyService(memberService, studyRepository);
+        
+        assertNotNull(studyService);
+    }
+}
+```
+
+2. `@Mock`
+- 애노테이션을 붙여서 Mock 객체를 생성할 수 있다.
+
+```java
+@Mock List mockedList;
+```
+
+```java
+class StudyServiceTest {
+
+    @Mock MemberService memberService;
+
+    @Mock StudyRepository studyRepository
+
+    @Test
+    void createStudyService() {
+        StudyService studyService = new StudyService(memberService, studyRepository);
+        assertNotNull(studyService);
+    }
+}
+```
+
+- 이 경우 Mock 객체는 null이다. `@Mock` 애노테이션만 붙여준다고 Mock 객체를 생성해주지 않기 때문이다.
+- `@ExtendWith(MockitoExtension.class)`를 붙여야 Mock 객체를 생성할 수 있다.
+
+```java
+@ExtendWith(MockitoExtension.class)
+class StudyServiceTest {
+
+    @Mock MemberService memberService;
+    @Mock StudyRepository studyRepository;
+
+    @Test
+    void createStudyService() {
+        StudyService studyService = new StudyService(memberService, studyRepository);
+        assertNotNull(studyService);
+    }
+}
+```
+
+## Mock 객체 주입 
+
+위 StudyServiceTest에서는 StudyService를 생성할 때 인자를 Mock 객체를 넘겨줘서 객체를 생성했다. 하지만 Mockito에서는 Mock 객체를 주입받으려는 대상에 `@InjectMocks`을 붙이면 자동으로 Mock 객체를 주입해준다.
+
+```java
+@ExtendWith(MockitoExtension.class)
+class StudyServiceTest {
+
+    @Mock MemberService memberService;
+    @Mock StudyRepository studyRepository;
+
+    @InjectMocks StudyService studyService;
+
+    @Test
+    void createStudyService() {
+        assertNotNull(studyService);
+    }
+}
+```
+
+## Mock 객체 Stubbing
+- **stubbing** : Mock 객체의 행동을 조작하는 것
+
+>[method stub](https://ko.wikipedia.org/wiki/%EB%A9%94%EC%86%8C%EB%93%9C_%EC%8A%A4%ED%85%81) : 다른 프로그래밍 기능을 대리하는 코드. stub은 기존 코드를 흉내내거나 아직 개발되지 않은 코드를 임시로 대치하는 역할을 수행한다.
+
+> Q. 언제 stub을 사용할까?
+> A. stub은 인터페이스가 정의는 되어있으나 아직 구현되지 않았을 때, 인터페이스의 실제 구현 대신으로 주로 사용된다.
+
+```java
+LinkedList mockedList = mock(LinkedList.class);
+
+//stubbing
+when(mockedList.get(0)).thenReturn("first");
+when(mockedList.get(1)).thenThrow(new RuntimeException());
+
+assertEquals("first", mockedList.get(0));
+assertThrows(RuntimeException.class, () -> { mockedList.get(1); });
+assertNull(mockedList.get(999));
+
+```
+- `thenReturn("first")` : mockedList.get(0)이 호출되었을 때 "first"를 리턴한다.
+- `thenThrow(new RuntimeException())` : mockedList.get(1)이 호출되었을 때 RuntimeExcepton이 발생한다.
+- mockedList.get(999)이 호출되었을 때 행동은 정의되지 않았다. => null 리턴
+
+```java
+@ExtendWith(MockitoExtension.class)
+class StudyServiceTest {
+
+    @Mock MemberService memberService;
+    @Mock StudyRepository studyRepository;
+
+    @InjectMocks StudyService studyService;
+
+    @Test
+    void createStudyService() {
+        Member member = new Member();
+        member.setId(1L);
+        member.setEmail("test@naver.com");
+
+        when(memberService.findById(1L)).thenReturn(Optional.of(member));
```

Comment:
제 기억으로는 stubbing을 정의할 때 인자는 argument matcher를 사용해야하는 걸로 알고 있는데요.
이렇게 되야하는 게 아닌가 싶었어요
```java
when(memberService.findById(eq(1L))).thenReturn(Optional.of(member));
```
작성해주신 예시를 봐서는 제대로 동작을 하나보네요..?

Reply:
넵 테스트했을 때는 문제 없이 동작했어요! 

궁금해서 찾아봤는데 경철님이 말씀하신 게 이 부분 같아요
> **Warning on argument matchers:**
>If you are using argument matchers, all arguments have to be provided by matchers.
```java
verify(mock).someMethod(anyInt(), anyString(), eq("third argument")); //above is correct - eq() is also an argument matcher

verify(mock).someMethod(anyInt(), anyString(), "third argument"); //above is incorrect - exception will be thrown because third argument is given without an argument matcher.
```
제가 이해하기로는 **stubbing을 할 때 인자에 argument matcher를 사용할거면 모든 인자에 사용해야 한다**는 것 같아요. 위 예시에서 stubbing을 해야 하는 메서드의 인자가 3개인데 이 때 2개가 argument matcher를 사용했는데 나머지 하나가 그냥 string이 들어가면 안된다. 이 의미죠.

예제 코드에서는 argument matcher를 사용 안 한 예시여서 동작하지 않았을까요??